src/main/java/com/ning/http/util/UTF8UrlEncoder.java

```diff
@@ -23,25 +23,178 @@
+            ALPHA[i] = true;
         }
         for (int i = 'A'; i <= 'Z'; ++i) {
-            SAFE_ASCII[i] = 1;
+            ALPHA[i] = true;
         }
+    }
+
+    private final static boolean[] DIGIT = new boolean[128];
+
+    static {
         for (int i = '0'; i <= '9'; ++i) {
-            SAFE_ASCII[i] = 1;
+            DIGIT[i] = true;
+        }
+    }
+
+    /**
+     * unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+     */
+    private final static boolean[] UNRESERVED = new boolean[128];
+    static {
+        for (int i = 0; i < 128; i++) {
+            UNRESERVED[i] = ALPHA[i] || DIGIT[i];
+        }
+        UNRESERVED['-'] = true;
+        UNRESERVED['.'] = true;
+        UNRESERVED['_'] = true;
+        UNRESERVED['~'] = true;
+    }
+
+    /**
+     *  scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+     */
+    private final static boolean[] SCHEME = new boolean[128];
+    static {
+        for(int i=0; i < 128; i++) {
+            SCHEME[i] = ALPHA[i] || DIGIT[i];
+        }
+        SCHEME['+'] = true;
+        SCHEME['-'] = true;
+        SCHEME['.'] = true;
+    }
+
+    /**
+     *  userinfo    = *( unreserved / pct-encoded / sub-delims / ":" )
+     */
+    private final static boolean[] USERINFO = new boolean[128];
+    static {
+        for(int i=0; i < 128; i++) {
+            USERINFO[i] = UNRESERVED[i] || SUB_DELIMS[i];
+        }
+        USERINFO[':'] = true;
+    }
+
+    /**
+     * host        = IP-literal / IPv4address / reg-name
+     * reg-name    = *( unreserved / pct-encoded / sub-delims )
+     */
+    private final static boolean[] HOSTPORT = new boolean[128];
+    static {
+        for(int i=0; i < 128; i++) {
+            HOSTPORT[i] = UNRESERVED[i] || SUB_DELIMS[i];
         }
-        SAFE_ASCII['-'] = 1;
-        SAFE_ASCII['.'] = 1;
-        SAFE_ASCII['_'] = 1;
-        SAFE_ASCII['~'] = 1;
+        HOSTPORT[':'] = true;
+    }
+
+    /**
+     * pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+     */
+    private final static boolean[] PCHAR = new boolean[128];
+    static {
+        for(int i=0; i < 128; i++) {
+            PCHAR[i] = UNRESERVED[i] || SUB_DELIMS[i];
+        }
+        PCHAR[':'] = true;
+        PCHAR['@'] = true;
+    }
+
+    /**
+     * query       = *( pchar / "/" / "?" )
+     */
+    private final static boolean[] QUERY = new boolean[128];
+    static {
+        System.arraycopy(PCHAR, 0, QUERY, 0, 128);
+        QUERY['/'] = true;
+        QUERY['?'] = true;
+    }
+
+    /**
+     * used to encode individual name and value parts of query
+     * =& are characters used to separate name and value pairs
+     * + needs to be encoded to differentiate from encoded spaces
+     */
+    private final static boolean[] QUERY_PARAM = new boolean[128];
+    static {
+        System.arraycopy(QUERY, 0, QUERY_PARAM, 0, 128);
+        QUERY_PARAM['='] = false;
+        QUERY_PARAM['+'] = false;
+        QUERY_PARAM['&'] = false;
```

Comment:
If you just checked by typing that into the query string, that's kinda like saying that typing "foo=bar&baz" will send as is. Yes, it will send that, but a server will see `foo` as just `bar`, and if you wanted `"bar&baz"` you'd have to URL-encode the ampersand. Same with ";" for some servers.